R/cat-wordwrap.R

```diff
@@ -0,0 +1,101 @@
+# Utility functions for performing word wrapping on streaming text. Our strategy
+# for dealing with streaming is to buffer the most recent word, assuming that
+# it's incomplete, and render it only when the next word break (space or
+# newline) is encountered.
+
+
+# Split a string into contiguous runs of non-spaces, " ", and newline.
+#
+# split_spaces("one  two    three") => c("one", "  ", "two", "    ", "three")
+split_spaces <- function(text) {
+  if (length(text) != 1 || !is.character(text)) {
+    stop("Input must be a character vector of length 1")
+  }
+
+  pattern <- "([^ \\n]+| +|\\n)"
+  result <- regmatches(text, gregexpr(pattern, text, perl = TRUE))[[1]]
+  result
+}
+
+cat_word_wrap_impl <- coro::generator(function(con = stdout()) {
+  CAT <- function(...) {
+    cat(..., file = con, sep = "")
+  }
+
+  pos_cursor <- 1
+  buffer <- ""
+
+  while (TRUE) {
+    input <- coro::yield()
```

Comment:
Oh I guess you're using coro here, but presumably this wouldn't be too hard to reimplement as a normal function and not a generator?

Reply:
The moment you use `chat$console()` or `chat$chat(echo=TRUE)` we need coro anyway. And coro has no imports other than rlang. I thought it was worth the dependency.